File: Java/ObserverPattern/src/com/company/Observable/WeatherStation.java
package com.company.Observable;

import com.company.Observers.IObserver;

import java.util.ArrayList;
import java.util.List;

public class WeatherStation implements IWeatherStation {

    private List<IObserver> observers;
    private int temperature;
    private int humidity;

    public WeatherStation() {
        this.observers = new ArrayList<>();
    }

    @Override
    public void registerObserver(IObserver o) {
        observers.add(o);
    }

    @Override
    public void unregisterObserver(IObserver o) {
        int observerIndex = observers.indexOf(o);
        if(observerIndex >= 0){
            observers.remove(observerIndex);
        }
    }

    @Override
    public void notifyObservers() {
        observers.forEach(observer -> observer.updateObserver(temperature, humidity));
    }

    public void weatherChanges(int temperature, int humidity){
        this.temperature = temperature;
        this.humidity = humidity;
        notifyObservers();
    }
}

File: Kotlin/ObserverPattern/src/observable/WeatherStation.kt
package observable

import observers.IObserver

/**
 *
 *   Author: Utkarsh Dayal @ descifrador17
 *   Date: 26-05-2021
 *   Time: 06:01 PM
 *
 */
class WeatherStation : IWeatherStation{

    private var observers = mutableListOf<IObserver>()

    private var temperature: Int = 0
    private var humidity: Int = 0


    override fun registerObserver(observer: IObserver) {
        observers.add(observer)
    }

    override fun unregisterObserver(observer: IObserver) {
        observers.remove(observer)
    }

    override fun notifyAllObservers() {
        observers.forEach { it.updateObserver(temperature, humidity) }
    }

    fun weatherChanges(temperature: Int, humidity: Int) {
        this.temperature = temperature
        this.humidity = humidity
        notifyAllObservers()
    }

}